### PR TITLE
Homomorphic XOR and NOT gates

### PR DESCRIPTION
**Pull request by hilder-vitor**
- Created homomorphic XOR and NOT gates.
- Created tests for these gates.
- Refactored tests to instantiate SchemeNTRU a single time because generating the key several times was taking too much time.

### CODE REVIEW

File: include/ntruhe.h [hunk 1]
```diff
@@ -151,19 +160,35 @@ class SchemeNTRU
 
     /**
      * Computes the AND gate of two given ciphertexts ct1 and ct2
-     * @param[out] ct_res encryptions of the outuput of the NAND gate
```
Comment (iliailia) [inline include/ntruhe.h:154]: Thanks for fixing these comments!

File: src/lwehe.cpp [hunk 2]
```diff
@@ -470,4 +470,27 @@ void SchemeLWE::or_gate(Ctxt_LWE& ct_res, const Ctxt_LWE& ct1, const Ctxt_LWE& c
...
     bootstrap(ct_res);
     //cout << "OR: " << float(clock()-start)/CLOCKS_PER_SEC << endl;
-}
\ No newline at end of file
+}
+
+
+void SchemeLWE::xor_gate(Ctxt_LWE& ct_res, const Ctxt_LWE& ct1, const Ctxt_LWE& ct2) const
+{
+    if (ct_res.a.size() != parLWE.n)
```
Comment (iliailia) [inline src/lwehe.cpp:478]: You can simplify here by writing 

```c++
ct_res = ct1 + ct2;
ct_res = ct_res + ct_res;
```

File: src/lwehe.cpp [hunk 3]
```diff
@@ -470,4 +470,27 @@ void SchemeLWE::or_gate(Ctxt_LWE& ct_res, const Ctxt_LWE& ct1, const Ctxt_LWE& c
...
+    for (size_t i = 0; i < parLWE.n; i++)
+        ct_res.a[i] = parLWE.mod_q_base(2*(ct1.a[i] + ct2.a[i])); // a = 2*(ct1.a + ct2.a)
+
+    ct_res.b = parLWE.mod_q_base(2*(ct1.b + ct2.b)); // b = 2*(ct1.b + ct2.b)
+    bootstrap(ct_res);
+}
+
+void SchemeLWE::not_gate(Ctxt_LWE& ct_res, const Ctxt_LWE& ct) const
+{
+    if (ct_res.a.size() != parLWE.n)
```
Comment (iliailia) [inline src/lwehe.cpp:489]: Here, you can simplify by writing only
```c++
ct_res = parLwe.delta_base - ct;
```

File: src/ntruhe.cpp [hunk 4]
```diff
@@ -282,4 +282,22 @@ void SchemeNTRU::or_gate(Ctxt_NTRU& ct_res, const Ctxt_NTRU& ct1, const Ctxt_NTR
     ct_res = ct_or_const - ct1 - ct2;
     bootstrap(ct_res);
     //cout << "OR: " << float(clock()-start)/CLOCKS_PER_SEC << endl;
-}
\ No newline at end of file
+}
+
+void SchemeNTRU::xor_gate(Ctxt_NTRU& ct_res, const Ctxt_NTRU& ct1, const Ctxt_NTRU& ct2) const
+{
+    if (ct_res.data.size() != parNTRU.n)
```
Comment (iliailia) [inline src/ntruhe.cpp:289]: You can simplify here by writing 

```c++
ct_res = ct1 + ct2;
ct_res = ct_res + ct_res;
```

File: test.cpp [hunk 5]
```diff
@@ -452,51 +496,85 @@ void test_lwehe_gate_helper(int in1, int in2, SchemeLWE& s, GateType g)
...
+
+    cout << "NOT GATE IS OK" << endl;
 }
 
+
 void test_lwehe_gate_composition_helper(SchemeLWE& s, GateType g)
 {
     float avg_time = 0.0;
-    int N_TESTS = 100;
+    int N_TESTS = 110;
```
Comment (iliailia) [inline test.cpp:577]: :) Why 110?

**Review by iliailia — CHANGES_REQUESTED** [review]
Thank you for the changes! I left a few comments where you can simplify the code; the rest looks great!

**Comment by hilder-vitor** [comment]
Thanks for the comments. Indeed, the code of those gates are simpler now. I also ran the tests and everything is still working fine, so I think you can merge it now.

**Review by iliailia — APPROVED** [review]
Thanks!

**Pull request merged**